java/java.openjdk.project/src/org/netbeans/modules/java/openjdk/project/RemotePlatformImpl.java

```diff
@@ -0,0 +1,136 @@
+import org.openide.filesystems.FileChangeAdapter;
+import org.openide.filesystems.FileChangeListener;
+import org.openide.filesystems.FileEvent;
+import org.openide.filesystems.FileObject;
+import org.openide.filesystems.FileUtil;
+import org.openide.util.ChangeSupport;
+import org.openide.util.Lookup;
+import org.openide.util.lookup.Lookups;
+
+/**
+ *
+ * @author lahvac
+ */
+public class RemotePlatformImpl implements RemotePlatform {
+
+    private static final Map<FileObject, RemotePlatformImpl> jdkRoot2Platform = new HashMap<>();
+
+    public static @NonNull RemotePlatformImpl getProvider(FileObject jdkRoot, ConfigurationImpl.ProviderImpl configurations) {
+        return jdkRoot2Platform.computeIfAbsent(jdkRoot, r -> new RemotePlatformImpl(configurations));
+    }
+
+    private final ChangeSupport cs = new ChangeSupport(this);
+    private final ConfigurationImpl.ProviderImpl configurations;
+    private File modules;
+    private FileChangeListener modulesListener = new FileChangeAdapter() {
+        @Override
+        public void fileChanged(FileEvent fe) {
+            cs.fireChange();
+        }
+    };
+
+    public RemotePlatformImpl(ConfigurationImpl.ProviderImpl configurations) {
+        this.configurations = configurations;
+        this.configurations.addPropertyChangeListener(new PropertyChangeListener() {
+            @Override public void propertyChange(PropertyChangeEvent evt) {
+                if (ProjectConfigurationProvider.PROP_CONFIGURATION_ACTIVE.equals(evt.getPropertyName()) ||
+                    evt.getPropertyName() == null) {
+                    listenOnModules();
+                    cs.fireChange();
+                }
+            }
+        });
+        listenOnModules();
+    }
+
+    private synchronized void listenOnModules() {
+        if (modules != null) {
+            FileUtil.removeFileChangeListener(modulesListener, modules);
+        }
+        modules = FileUtil.normalizeFile(new File(configurations.getActiveConfiguration().getLocation(),
+                                         "images/jdk/lib/modules".replace("/", System.getProperty("file.separator"))));
+        FileUtil.addFileChangeListener(modulesListener, modules);
+    }
+
+    @Override
+    public String getJavaCommand() {
+        return new File(configurations.getActiveConfiguration().getLocation(),
+                        "images/jdk/bin/java".replace("/", System.getProperty("file.separator")))
+               .getAbsolutePath();
+    }
+
+    @Override
+    public List<String> getJavaArguments() {
+        return Collections.emptyList();
+    }
+
+    @Override
+    public void addChangeListener(ChangeListener l) {
+        cs.addChangeListener(l);
+    }
+
+    @Override
+    public void removeChangeListener(ChangeListener l) {
+        cs.removeChangeListener(l);
+    }
+
+    public static Lookup createProvider(FileObject jdkRoot, JDKProject project) {
+        Provider provider = new Provider() {
```

Comment:
Jackpot:
warning: This anonymous inner class creation can be turned into a lambda expression.